website/server/libs/payments.js

```diff
@@ -44,6 +45,158 @@ function _dateDiff (earlyDate, lateDate) {
+      key: 'group_plan_auto',
+    },
+    customerId: 'group-plan',
+    paymentMethod: 'Group Plan',
+    headers: {},
+  };
+
+  let plan = {
+    planId: 'group_plan_auto',
+    customerId: 'group-plan',
+    dateUpdated: new Date(),
+    gemsBought: 0,
+    paymentMethod: 'groupPlan',
+    extraMonths: 0,
+    dateTerminated: null,
+    lastBillingDate: null,
+    dateCreated: new Date(),
+    mysteryItems: [],
+    consecutive: {
+      trinkets: 0,
+      offset: 0,
+      gemCapExtra: 0,
+    },
+  };
+
+  if (member.isSubscribed()) {
+    let memberPlan = member.purchased.plan;
+    let customerHasCancelledGroupPlan = memberPlan.dateTerminated && memberPlan.customerId === this.constants.GROUP_PLAN_CUSTOMER_ID && moment().isBefore(memberPlan.dateTerminated);
+    if (customerIdsToIgnore.indexOf(memberPlan.customerId) !== -1 && !customerHasCancelledGroupPlan) return;
+
+    await member.cancelSubscription();
+
+    let today = new Date();
+    plan = _.clone(member.purchased.plan.toObject());
+    let extraMonths = Number(plan.extraMonths);
+    if (plan.dateTerminated) extraMonths += _dateDiff(today, plan.dateTerminated);
+
+    _(plan).merge({ // override with these values
+      planId: 'group_plan_auto',
+      customerId: 'group-plan',
+      dateUpdated: today,
+      paymentMethod: 'groupPlan',
+      extraMonths,
+      dateTerminated: null,
+      lastBillingDate: null,
+      owner: member._id,
+    }).defaults({ // allow non-override if a plan was previously used
+      gemsBought: 0,
+      dateCreated: today,
+      mysteryItems: [],
+    }).value();
+  }
+
+  member.purchased.plan = plan;
+
+  data.user = member;
+  await this.createSubscription(data);
+
+  let leader = await User.findById(group.leader).exec();
+  txnEmail(data.user, 'group-member-joining', [
+    {name: 'LEADER', content: leader.profile.name},
+    {name: 'GROUP_NAME', content: group.name},
+  ]);
+};
+
+
+/**
+ * Cancels subscriptions of members of a group
+ *
+ * @param  group  The Group Model that is cancelling a group plan
+ *
+ * @return undefined
+ */
+api.cancelGroupUsersSubscription = async function cancelGroupUsersSubscription (group) {
+  let members;
+  if (group.type === 'guild') {
+    members = await User.find({guilds: group._id}).select('_id guilds purchased').exec();
+  } else {
+    members = await User.find({'party._id': group._id}).select('_id guilds purchased').exec();
+  }
+
+  let promises = members.map((member) => {
+    return this.cancelGroupSubscriptionForUser(member, group);
+  });
+
+  await Promise.all(promises);
+};
+
+api.cancelGroupSubscriptionForUser = async function cancelGroupSubscriptionForUser (user, group) {
+  if (user.purchased.plan.customerId !== this.constants.GROUP_PLAN_CUSTOMER_ID) return;
+
+  let userGuilds = _.clone(user.guilds);
+  userGuilds.push('party');
+
+  let index = userGuilds.indexOf(group._id);
+  userGuilds.splice(index, 1);
+
+  let groupPlansQuery = {
+    type: {$in: ['guild', 'party']},
```

Comment:
oh I see, can we change the name of userGuilds to something else, otherwise it's super confusing